### ISSUE TITLE
Broken build if a dependency's package.json uses the production export condition

### ISSUE THREAD

**Issue by marbemac** [nitrojs/nitro#2923]
### Environment

- Node v18.19.0
- Nitro current v2 branch

### Reproduction

I've updated one of the nitro examples here to surface the bug: https://github.com/marbemac/nitro/commit/9ecc35e38226aa0757f8f6ffa166f3ec74d29cca

In this example I've added the `@zedux/react` package, which uses the production conditional in their package.json exports.

1. switch to that commit or copy the changes
2. `pnpm i`
3. `cd examples/nano-jsx`
4. `pnpm build`
5. `node .output/server/index.mjs`
6. open server url in browser, see error in console

<img width="864" alt="Screenshot 2024-12-05 at 4 22 12 PM" src="https://github.com/user-attachments/assets/56ffff77-faab-476d-9a28-fb768deccce4">

### Describe the bug

In the example described in the reproduction.

Here is what the original package.json looks like (from node_modules):

```json
{
  "name": "@zedux/react",
  "exports": {
    ".": {
      "import": "./dist/esm/index.js",
      "require": "./dist/cjs/index.js",
      "production": {
        "import": "./dist/zedux-react.es.min.js",
        "require": "./dist/zedux-react.umd.min.js",
        "default": "./dist/zedux-react.umd.min.js"
      },
      "default": "./dist/cjs/index.js"
    }
  }
}
```

And here is what it looks like in the build `.output` folder's node_modules:

```json
{
  "name": "@zedux/react",
  "exports": {
    ".": {
      "import": "./dist/zedux-react.es.min.js",
      "require": "./dist/zedux-react.umd.min.js",
      "production": {
        "import": "./dist/zedux-react.es.min.js",
        "require": "./dist/zedux-react.umd.min.js",
        "default": "./dist/zedux-react.umd.min.js"
      },
      "default": "./dist/zedux-react.umd.min.js"
    }
  }
}
```

The exports were correctly re-written to all match the production export (tests cover this here -> https://github.com/nitrojs/nitro/blob/v2/test/unit/externals.test.ts). However, note in the screenshot below that only the `dist/esm` folder is included in the build output's node_modules, perhaps because Nitro ignored the production conditional, and moved files over according to the original "import" export (before it is rewritten)? No idea as I'm not familiar with the Nitro internals, but maybe in the `buildEnd` step in the `externals.ts` rollup plugin.

This results in a mismatch, and runtime error when importing the affected package(s).

<img width="384" alt="Screenshot 2024-12-05 at 3 50 02 PM" src="https://github.com/user-attachments/assets/809b5c4e-55bb-49c8-9082-54d52e5b27b7">

### Additional context

I believe this will affect any projects with dependencies that make use of the production conditional in their package.json exports. Perhaps other conditionals, not sure. Might be related to #2242.

**Comment by pi0** [comment]
This looks like a bug if happening. Nitro tracer automatically should rewrite export conditions (see https://github.com/nitrojs/nitro/issues/844 for reference) 

Regardless, a workaround is to add particular dependencies to `externals.traceInclude: []` array this way nitro traces all files.

I will check it closely soon (if not too much trouble can you please create a stackblitz also? that would be super helpful for me to clone and try against published nitro version)

**Comment by marbemac** [comment]
Been investigating for the last 30 minutes and I the issue might actually be in https://github.com/vercel/nft. Seeing if I can reproduce with a test over there and will report back.

**Comment by pi0** [comment]
Thanks for follwup update. I will investigate better but quick look, `production` could be better the first condition in the lib. (consider ESM resolution order is based on the actual package `exports` keys order not conditions of resolution) so it should have been something like this for proper behavior:

```js
{
  "name": "@zedux/react",
  "exports": {
    ".": {
      "production": {
        "import": "./dist/zedux-react.es.min.js",
        "require": "./dist/zedux-react.umd.min.js",
        "default": "./dist/zedux-react.umd.min.js"
      },
      "import": "./dist/esm/index.js",
      "require": "./dist/cjs/index.js",
      "default": "./dist/cjs/index.js"
    }
  }
}
```

**Comment by marbemac** [comment]
Good tip on the traceInclude! I'll try it as a stop gap.

Does indeed look like NFT is picking whichever export key matches first. That combined with how NFT implicitly includes "default", "require", and "import" conditions means that they will always be picked if defined earlier in the exports object. https://github.com/vercel/nft/blob/main/src/resolve-dependency.ts#L166-L172

Is that how the node resolver works? Might be - and if so perhaps the issue is the implicit conditions that are always added by NFT?

**Comment by pi0** [comment]
> Is that how the node resolver works?

Yes, exports keys order matters for ESM resolution algorithm in general (this is also why `typescript` has to always go first for example)

**Comment by marbemac** [comment]
Yeah just verified, ok well then looks like it's an issue in the downstream dependency, as you suggested, thanks for your help!

**Comment by marbemac** [comment]
@pi0 thinking on it a bit more, would it make sense for nitro to adhere to the same node resolver behavior when re-writing the exports package.json field for external deps? Right now it seems to always attempt to use production, regardless of position in the exports: https://github.com/nitrojs/nitro/blob/v2/src/rollup/plugins/externals.ts#L485

Then again, not a huge deal I suppose, and might need to be considered a breaking change to change this behavior.

**Comment by pi0** [comment]
Yes you are right, nitro could rewrite to first matched condition instead of production hardcoded. Might be little tricky but doable.

**Comment by DannyVogel** [comment]
Encountered the same issue with @google/genai.

Error: Error [ERR_MODULE_NOT_FOUND]: Cannot find module '@google/genai/dist/node/index.mjs'

Cause: The package has conditional exports for node and web. Nitro only copied dist/web/ to .output/server/node_modules/, but at runtime Node.js resolved to dist/node/index.mjs which didn't exist.

  Workaround:
  //nitro.config.ts
```
  export default defineNitroConfig({
    externals: {
      inline: ['@google/genai'],
    },
  })
```

**Comment by fabis94** [comment]
Also hit the same issue just now. Is there any fix planned for this? If this is not the right package, where should a new issue report should be posted?
